engine/runtime/src/main/scala/org/enso/compiler/Cache.scala

```diff
@@ -0,0 +1,463 @@
+    sourceDigest: String,
+    blobDigest: String,
+    entry: T
+  ): Array[Byte]
+
+  /** Gets the path to the cache data within the `cacheRoot`.
+    *
+    * @param cacheRoot the root of the cache for this entry
+    * @return the name of the data file for this entry's cache
+    */
+  private def getCacheDataPath(cacheRoot: TruffleFile): TruffleFile = {
+    cacheRoot.resolve(cacheFileName(dataSuffix))
+  }
+
+  private def getCacheMetadataPath(cacheRoot: TruffleFile): TruffleFile =
+    cacheRoot.resolve(cacheFileName(metadataSuffix))
+
+  /** Computes the cache file name for a given extension.
+    *
+    * @param suffix the extension
+    * @return the cache file name with the provided `ext`
+    */
+  private def cacheFileName(suffix: String): String = {
+    s"$entryName$suffix"
+  }
+
+  protected def stringRepr: String
+
+  protected def entryName: String
+
+  protected def dataSuffix: String
+
+  protected def metadataSuffix: String
+
+  protected def messageDigest(): MessageDigest = {
+    new SHA3.Digest224()
+  }
+
+  /** Deletes the cache for this data in the provided `cacheRoot`.
+    *
+    * @param cacheRoot the root of the cache to delete
+    * @param logger    a logger
+    */
+  private def invalidateCache(
+    cacheRoot: TruffleFile
+  )(implicit logger: TruffleLogger): Unit = {
+    val metadataFile = getCacheMetadataPath(cacheRoot)
+    val dataFile     = getCacheDataPath(cacheRoot)
+
+    def doDeleteAt(file: TruffleFile): Unit = {
+      try {
+        if (file.exists()) {
+          if (file.isWritable) {
+            file.delete()
+            logger.log(
+              logLevel,
+              s"Invalidated the cache at [${file.toMaskedPath.applyMasking()}]."
+            )
+          } else {
+            logger.log(
+              logLevel,
+              s"Cannot invalidate the cache at " +
+              s"[${file.toMaskedPath.applyMasking()}]. " +
+              s"Cache location not writable."
+            )
+          }
+        }
+      } catch {
+        case _: NoSuchFileException =>
+        // If it doesn't exist, our work has already been done for us!
+        case _: java.nio.file.DirectoryNotEmptyException | _: IOException |
+            _: SecurityException =>
+          logger.log(
+            logLevel,
+            s"Unable to delete the cache at " +
+            s"[${cacheRoot.toMaskedPath.applyMasking()}]."
+          )
+      }
+    }
+
+    doDeleteAt(metadataFile)
+    doDeleteAt(dataFile)
+  }
+
+  /** Invalidates all caches associated with this cache.
+    *
+    * @param context the langage context in which loading is taking place
+    */
+  def invalidate(context: EnsoContext): Unit = {
+    this.synchronized {
+      implicit val logger: TruffleLogger = context.getLogger(this.getClass)
+      getCacheRoots(context).foreach { roots =>
+        invalidateCache(roots.globalCacheRoot)
+        invalidateCache(roots.localCacheRoot)
+      }
+    }
+  }
+
+}
+
+object Cache {
```

Comment:
Having a shared access to caches is a great move forward. I would, however, prefer if the new code was written in Java rather than Scala. If this was a new code (as I realized later)...

Reply:
This was a big one but I did it just for you :)